pkg/termination/termination.go

```diff
@@ -0,0 +1,172 @@
+	namespace    string
+	log          logr.Logger
+}
+
+// Run starts the handler and runs the termination logic
+func (h *handler) Run(stop <-chan struct{}) error {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	errs := make(chan error, 1)
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+		errs <- h.run(ctx)
+	}()
+
+	select {
+	case <-stop:
+		cancel()
+		// Wait for run to stop
+		wg.Wait()
+		return nil
+	case err := <-errs:
+		cancel()
+		return err
+	}
+}
+
+func (h *handler) run(ctx context.Context) error {
+	machine, err := h.getMachineForNode(ctx)
+	if err != nil {
+		return fmt.Errorf("error fetching machine for node (%q): %w", h.nodeName, err)
+	}
+
+	logger := h.log.WithValues("machine", machine.Name)
+	logger.V(1).Info("Monitoring node for machine")
+
+	if err := wait.PollImmediateUntil(h.pollInterval, func() (bool, error) {
+		req, err := http.NewRequest("GET", h.pollURL.String(), nil)
+		if err != nil {
+			return false, fmt.Errorf("could not create request %q: %w", h.pollURL.String(), err)
+		}
+
+		req.Header.Add("Metadata", "true")
+
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			return false, fmt.Errorf("could not get URL %q: %w", h.pollURL.String(), err)
+		}
+
+		bodyBytes, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return false, fmt.Errorf("failed to read responce body: %w", err)
+		}
+
+		s := scheduledEvents{}
+		err = json.Unmarshal(bodyBytes, &s)
+		if err != nil {
+			return false, fmt.Errorf("failed to unmarshal responce body: %w", err)
+		}
+
+		for _, event := range s.Events {
+			if event.EventType == preemptEventType {
+				// Instance marked for termination
+				return true, nil
+			}
+		}
+
+		// Instance not terminated yet
+		logger.V(2).Info("Instance not marked for termination")
+		return false, nil
+	}, ctx.Done()); err != nil {
+		return fmt.Errorf("error polling termination endpoint: %w", err)
+	}
+
+	// Will only get here if the termination endpoint returned FALSE
+	logger.V(1).Info("Instance marked for termination, deleting Machine")
+	if err := h.client.Delete(ctx, machine); err != nil {
+		return fmt.Errorf("error deleting machine: %w", err)
+	}
+
+	return nil
+}
+
+// getMachineForNodeName finds the Machine associated with the Node name given
+func (h *handler) getMachineForNode(ctx context.Context) (*machinev1.Machine, error) {
+	machineList := &machinev1.MachineList{}
+	err := h.client.List(ctx, machineList, client.InNamespace(h.namespace))
+	if err != nil {
+		return nil, fmt.Errorf("error listing machines: %w", err)
+	}
+
+	for _, machine := range machineList.Items {
+		if machine.Status.NodeRef != nil && machine.Status.NodeRef.Name == h.nodeName {
+			return &machine, nil
+		}
+	}
+
+	return nil, fmt.Errorf("machine not found for node %q", h.nodeName)
+}
+
+const preemptEventType = "Preempt"
+
+type scheduledEvents struct {
+	Events []events `json:"Events"`
+}
+
+type events struct {
+	EventType string `json:"EventType"`
+}
```

Comment:
If there's any documentation that explains the format of the metadata response, it would be good to link here I think